src/Auth0Client.ts

```diff
@@ -493,4 +488,89 @@ export default class Auth0Client {
     const url = this._url(`/v2/logout?${createQueryParams(logoutOptions)}`);
     window.location.assign(`${url}${federatedQuery}`);
   }
+
+  private async _getTokenFromIFrame(
+    options: GetTokenSilentlyOptions
+  ): Promise<any> {
+    const stateIn = encodeState(createRandomString());
+    const nonceIn = createRandomString();
+    const code_verifier = createRandomString();
+    const code_challengeBuffer = await sha256(code_verifier);
+    const code_challenge = bufferToBase64UrlEncoded(code_challengeBuffer);
+
+    const authorizeOptions = {
+      audience: options.audience,
+      scope: options.scope
+    };
+
+    const params = this._getParams(
+      authorizeOptions,
+      stateIn,
+      nonceIn,
+      code_challenge,
+      this.options.redirect_uri || window.location.origin
+    );
+
+    const url = this._authorizeUrl({
+      ...params,
+      prompt: 'none',
+      response_mode: 'web_message'
+    });
+
+    const codeResult = await runIframe(url, this.domainUrl);
+
+    if (stateIn !== codeResult.state) {
+      throw new Error('Invalid state');
+    }
+
+    const tokenResult = await oauthToken({
+      baseUrl: this.domainUrl,
+      audience: options.audience || this.options.audience,
+      client_id: this.options.client_id,
+      code_verifier,
+      code: codeResult.code,
+      grant_type: 'authorization_code'
+    } as OAuthTokenOptions);
+
+    const decodedToken = this._verifyIdToken(tokenResult.id_token, nonceIn);
+
+    return {
+      ...tokenResult,
+      decodedToken,
+      scope: params.scope,
+      audience: params.audience || 'default'
+    };
+  }
+
+  private async _getTokenUsingRefreshToken(
+    options: GetTokenSilentlyOptions
+  ): Promise<any> {
+    const cache = this.cache.get({
+      scope: options.scope,
+      audience: options.audience || 'default'
+    });
+
+    if (!cache || !cache.refresh_token) {
+      throw new GenericError(
+        'missing_refresh_token',
+        'No refresh token is available to fetch a new access token. The user should be reauthenticated.'
+      );
+    }
+
+    const tokenResult = await oauthToken({
+      baseUrl: this.domainUrl,
+      client_id: this.options.client_id,
+      grant_type: 'refresh_token',
+      refresh_token: cache.refresh_token
+    } as RefreshTokenOptions);
+
+    const decodedToken = this._verifyIdToken(tokenResult.id_token);
```

Comment:
Just to make sure ... does this handle the case where there is no ID token? If scopes were only OAuth2 and not OIDC?

Reply:
Yes, the ID token ultimately ends up in the `verify` method of jwt.js, which checks that is has a value.

https://github.com/auth0/auth0-spa-js/blob/8ceb626cc01cf85b845a763bfffc789a2ecce250/src/jwt.ts#L63